KAFKA-12281: Deprecate BrokerNotFoundException

karuturi opened the pull request:
Implements KIP-1195.

BrokerNotFoundException exception is unused since 2.8  Marking it
deprecated so that it can be removed in next major release.

Reviewers: Matthias J. Sax <matthias@confluent.io>

karuturi committed:
KAFKA-12281 deprecate BrokerNotFoundException

This exception is unused since 2.8. Marking it deprecated so that it can
be removed in next major release

mjsax commented on streams/src/main/java/org/apache/kafka/streams/errors/BrokerNotFoundException.java:
```diff
@@ -23,6 +23,7 @@
  * @see org.apache.kafka.streams.StreamsConfig
  */
 @SuppressWarnings("unused")
+@Deprecated(since = "4.2", forRemoval = true)
```
We should update the JavaDocs with `@deprecated` tag:
```
@deprecated Since 4.2 and should not be used any longer.
```

karuturi replied:
updated javadoc

mjsax commented on streams/src/main/java/org/apache/kafka/streams/errors/BrokerNotFoundException.java:
```diff
@@ -23,6 +23,7 @@
  * @see org.apache.kafka.streams.StreamsConfig
  */
 @SuppressWarnings("unused")
+@Deprecated(since = "4.2", forRemoval = true)
```
```suggestion
@Deprecated
```

mjsax replied:
We usually don't add metadata to the tag, but use JavaDocs for this.

karuturi replied:
ok, I followed some of the existing usages like `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter` 
Should I remove it?

mjsax replied:
Interesting. `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter` is an internal class though (note, it's in package `internals`), and it not user facing.

We usually just use plain `@Deprecated` (for example, compare `JoinWindows`).

karuturi replied:
used plain `@Deprecated` annotation

mjsax reviewed (commented):
Need to wait until the KIP is approved before, this PR can get merged.

karuturi committed:
KAFKA-12281 updated javadoc with @deprecated for BrokerNotFoundException

mjsax commented on streams/src/main/java/org/apache/kafka/streams/errors/BrokerNotFoundException.java:
```diff
@@ -17,12 +17,14 @@
 package org.apache.kafka.streams.errors;
 
 /**
+ * @deprecated since 4.2 and should not be used any longer.
```
nit: this line should on at the end of the JavaDoc comment. Not the top :)

karuturi replied:
moved to the end of javadoc

karuturi committed:
KAFKA-12281 addressed review comments

Following convention, moved deprecated tag to the end of javadoc and
removed version tags in @Deprecated annotation

karuturi commented:
> Need to wait until the KIP is approved before, this PR can get merged.

@mjsax KIP is approved now.

mjsax commented:
Thanks for the KIP and PR! Merged to `trunk`.

mjsax commented:
Filed https://issues.apache.org/jira/browse/KAFKA-19557 as follow up for AK 5.0.0 release.
